### PR TITLE
Refactor buffer mapping async algorithm

### PR DESCRIPTION
**Pull request by JiangYizhou**
1.Take care of upper limit of the buffer pool. If buffer and size is enough in the buffer pool, create a new buffer. If size reach the limit of the buffer pool, force wait for the buffer on mapping.
2.Use a ring buffer to upload data for all of the constant buffers. Return the buffer and its offset when allocate the buffer from the pool. Get the last one from the MappedBufferQueue and check if the ring buffer is full. If the buffer can hold extra size space, use the last one directly.
3.Return null if size reach the limit or no available buffer, this means small bubbles in some of the ring buffers and we haven't deal with the problem now.

### CODE REVIEW

File: src/aquarium-optimized/dawn/BufferManagerDawn.cpp [hunk 1]
```diff
@@ -128,26 +133,59 @@ RingBufferDawn *BufferManagerDawn::allocate(size_t size, bool sync)
 
         if (ringBuffer == nullptr)
         {
-            if (mUsedSize + size > mBufferPoolSize)
+            if (mCount < BUFFER_MAX_COUNT)
+            {
+                mUsedSize += size;
+                std::cout << "allocate new buffer" << mCount << std::endl;
```
Comment (hujiajie) [inline src/aquarium-optimized/dawn/BufferManagerDawn.cpp]: I'm not sure if there's any `std::cout` in the hot paths. If yes, please ensure logging is disabled in release build.

File: src/aquarium-optimized/BufferManager.h [hunk 2]
```diff
@@ -11,7 +11,9 @@
 #include <vector>
 #include "Context.h"
 
-constexpr size_t BUFFER_POOL_MAX_SIZE = 409600000;
+constexpr size_t BUFFER_POOL_MAX_SIZE    = 409600000;
+constexpr size_t BUFFER_MAX_COUNT        = 10;
+constexpr size_t BUFFER_PER_ALLOCAT_SIZE = BUFFER_POOL_MAX_SIZE / BUFFER_MAX_COUNT;
```
Comment (hujiajie) [inline src/aquarium-optimized/BufferManager.h]: Is `ALLOCAT` intentional or a typo?

File: src/aquarium-optimized/dawn/BufferManagerDawn.cpp [hunk 3]
```diff
@@ -88,36 +90,39 @@ BufferManagerDawn::~BufferManagerDawn()
...
         while (!mMappedBufferList.empty())
         {
             ringBuffer = static_cast<RingBufferDawn *>(mMappedBufferList.front());
-            mMappedBufferList.pop();
-            if (ringBuffer->getSize() < size)
+            if (ringBuffer->getAvailableSize() < size)
             {
-                mUsedSize += ringBuffer->getSize();
-                ringBuffer->destory();
+                mMappedBufferList.pop();
```
Comment (hujiajie) [inline src/aquarium-optimized/dawn/BufferManagerDawn.cpp:124]: I haven't understood the whole picture here. Why to pop the buffer when it's almost full? What will happen to the data in this buffer?

File: src/aquarium-optimized/dawn/ContextDawn.cpp [hunk 4]
```diff
@@ -605,14 +604,16 @@ void ContextDawn::initGeneralResources(Aquarium* aquarium)
...
 
 void ContextDawn::updateWorldlUniforms(Aquarium* aquarium)
 {
-    setBufferData(mLightWorldPositionBuffer, 0, sizeof(LightWorldPositionUniform),
-                  &aquarium->lightWorldPositionUniform);
+    updateBufferData(mLightWorldPositionBuffer, &aquarium->lightWorldPositionUniform,
+                     CalcConstantBufferByteSize(sizeof(LightWorldPositionUniform)));
+    // setBufferData(mLightWorldPositionBuffer, 0,
+    //            CalcConstantBufferByteSize(sizeof(LightWorldPositionUniform)),
+    //           &aquarium->lightWorldPositionUniform);
```
Comment (hujiajie) [inline src/aquarium-optimized/dawn/ContextDawn.cpp]: Are these comments still useful?

File: src/aquarium-optimized/dawn/FishModelInstancedDrawDawn.cpp [hunk 5]
```diff
@@ -175,6 +175,10 @@ void FishModelInstancedDrawDawn::init()
                                 {4, mNormalTexture->getTextureView()}});
     }
 
+    // mContextDawn->updateBufferData(mLightFactorBuffer,
+    //                            &mLightFactorUniforms, sizeof(LightFactorUniforms));
+    // mContextDawn->updateBufferData(mFishVertexBuffer,
+    //                            &mFishVertexUniforms, sizeof(FishVertexUniforms));
```
Comment (hujiajie) [inline src/aquarium-optimized/dawn/FishModelInstancedDrawDawn.cpp]: Is this unfinished work?

File: src/aquarium-optimized/dawn/BufferManagerDawn.cpp (hunk 3 above)
Comment (JiangYizhou) [reply src/aquarium-optimized/dawn/BufferManagerDawn.cpp:124]: MappedBufferLists contains ring buffers have available space. If the first buffer is full, pop it(not destroy) and add it to enquedBufferList, and then use the next empty buffer. Commit all of the staging buffer in enquedBufferList to update all of the data of a frame, and then remap the buffers, and they will be recycled to MappedBufferList again.

File: src/aquarium-optimized/dawn/BufferManagerDawn.cpp (hunk 1 above)
Comment (JiangYizhou) [reply src/aquarium-optimized/dawn/BufferManagerDawn.cpp]: I think so, just want to show how many buffers are allocated here.

File: src/aquarium-optimized/dawn/BufferManagerDawn.cpp (hunk 3 above)
Comment (hujiajie) [reply src/aquarium-optimized/dawn/BufferManagerDawn.cpp:124]: I see, thanks

File: src/aquarium-optimized/dawn/FishModelInstancedDrawDawn.cpp (hunk 5 above)
Comment (JiangYizhou) [reply src/aquarium-optimized/dawn/FishModelInstancedDrawDawn.cpp]: Thanks, I will remove the comments. setBufferData is implemented by buffer mapping sync method in Aquarium while updateBufferData is buffer mapping async method if  the toggle is enabled. And sync method is used in initialization and async method is used to update data for all constant buffers for every frame.

**Pull request merged**